### ISSUE TITLE
❓Implicit Grant workflow

### ISSUE THREAD

**Issue by alexdilley** [feathersjs/authentication#687]
Almost certainly a n00bish question, for which I apologise, but:

What would be the sense/logic behind authenticating with a JWT to then be issued with a JWT?

i.e. the documented pattern of:

```js
app.service('authentication').hooks({
  before: {
    create: [
      auth.hooks.authenticate('jwt'), // verify the request's token
      // ...on success, emit the `login` event
    ],
  },
});
```

The workflow being that I have authenticated with a token provider (Auth0, for example) and possess a token with an `audience` suitable for authenticating with my Feathers API (signed with a secret known by both parties) – a la OAuth Implicit Grant. I then authenticate a socket connection with this token by calling the authentication service's `create` endpoint (using feathersjs/authentication-client), only to then have this token effectively superseded by one issued in the response from the server.

Is there a recommended way to handle service authentication (over a socket, in my case) from a SPA that uses an Implicit Grant?

**Comment by alexdilley** [comment]
(Partly related issue: #169.)

Is one option to implement something like the following for this use-case?

```js
import auth from '@feathersjs/authentication';
import { keep } from 'feathers-hooks-common';

// ...

app.service('authentication').hooks({
  before: {
    create: [
      // Whilst this [also] exists as an app-wide hook, because this endpoint
      // effectively authenticates a socket for the duration of its
      // connection, its wise to explicitly authenticate here _just in case_
      // the global hook is ever bypassed.
      auth.hooks.authenticate('jwt'),
    ],
  },
  after: {
    create: [
      // Return an empty response (discard the token) by `keep`ing nothing:
      // our token provider for this app is Auth0 -- we don't want a user to
      // possess a token with potentially mismatched access rights.
      keep(),
    ],
  },
});

// ...

app.hooks({
  before: {
    all: [
      // Endpoints are only accessible to users with a valid token.
      auth.hooks.authenticate('jwt'),
    ],
  },
})
```

**Comment by alexdilley** [comment]
Better `after` hook (nod to feathersjs/authentication-jwt#61):

```js
app.service('authentication').hooks({
  after: {
    create: [
      // Echo the token supplied when authenticating, ignoring the JWT created.
      ctx => {
        if (ctx.data && ctx.data.strategy === 'jwt') {
          ctx.result.accessToken = ctx.data.accessToken;
        }
        return ctx;
      },
    ],
  },
},
```

**Comment by daffl** [comment]
I have implemented this before using a [custom authentication strategy](https://github.com/feathersjs-ecosystem/feathers-authentication-custom) that verifies the token with the remote service (so you use `authenticate('auth0-jwt')`). Then you can still use the last `after` hook to make sure a user keeps getting the same one back.

**Comment by alexdilley** [comment]
Since the server is trusted, it could use the signing secret to verify the token via passport-jwt strategy, save communication overhead with the remote service?

**Comment by daffl** [comment]
Yep. Or cache it once it has been verified remotely until it expires (assuming you can't revoke them).

**Comment by alexdilley** [comment]
Thanks for your help, @daffl. I'll close the issue.
